### PR TITLE
SHAKE (SHA3 related) big endian fix

### PR DESCRIPTION
**Pull request by karel-m**
BIG ENDIAN strikes again.

A good news is that this is the only thing from all recent additions that failed on a BE machine.

### CODE REVIEW

**Comment by sjaeckel** [comment]
FYI I just tried the patch on big endian PPC64 qemu and the test fails with or without your patch...

I'm also getting this
```
src/hashes/sha3.c: In function ‘sha3_shake_done’:
src/hashes/sha3.c:273:19: warning: declaration of ‘i’ shadows a previous local [-Wshadow]
          unsigned i;
                   ^
src/hashes/sha3.c:259:18: note: shadowed declaration is here
    unsigned long i = 0;
```

**Comment by karel-m** [comment]
I have run my tests on HP-UX. Unfortunately it were not libtomcrypt's tests as I am not able to convince the HP_UX's make to consume our makefiles (not even makefile.unix). I'll try to investigate more.

**Comment by sjaeckel** [comment]
I liked 150fb6bbdc930dafe95620c2abf099e1abea5386 better, but nevermind (I also think it was missing a reset of `i`) :)

Did you investigate that patch yourself? Or did you find it somewhere else?

**Comment by karel-m** [comment]
> Did you investigate that patch yourself? Or did you find it somewhere else?

The patch is basically the same stuff that we already use in `sha3_done` - https://github.com/libtom/libtomcrypt/blob/develop/src/hashes/sha3.c#L241 It was only missing in `sha3_shake_done`. The original code is here - https://github.com/brainhub/SHA3IUF/blob/master/sha3.c#L297 (it was just adopted to our macros and `ulong*` types).

So the latest https://github.com/libtom/libtomcrypt/pull/205/commits/3f01f2ca22b4416d0f6eafb8c9e3f7d8bfed70f2 fails on your PPC64@qemu as well, right?

**Comment by karel-m** [comment]
I'll check it in the evening, gotta do some $work.

**Comment by sjaeckel** [comment]
```
# ./test
build == git-1.17-795-gcb66e08
...
cipher_hash_test....SHAKE256('') failed
Algorithm failed test vectors. (5) - sha3_shake
cipher_hash_test.c:34:sha3_shake_test()
# uname -m
ppc64
```
FYI I patched the build to include the git revision

**Comment by karel-m** [comment]
The latest version works fine for me on `AIX @ PowerPC` and `HP-UX @ PA-RISC`.

File: src/hashes/sha3.c [hunk 1]
```diff
@@ -231,33 +231,27 @@ int sha3_process(hash_state *md, const unsigned char *in, unsigned long inlen)
...
    return CRYPT_OK;
 }
 
 int sha3_shake_done(hash_state *md, unsigned char *out, unsigned long outlen)
 {
-   unsigned long i = 0;
-   /* sha3_shake_done can be called many times */
+   /* IMPORTANT NOTE: sha3_shake_done can be called many times */
+   unsigned long idx;
+   unsigned i;
```
Comment (sjaeckel) [inline src/hashes/sha3.c]: only 1 variable required

File: src/hashes/sha3.c [hunk 2]
```diff
@@ -268,16 +262,20 @@ int sha3_shake_done(hash_state *md, unsigned char *out, unsigned long outlen)
       md->sha3.s[md->sha3.word_index] ^= (md->sha3.saved ^ (CONST64(0x1F) << (md->sha3.byte_index * 8)));
       md->sha3.s[SHA3_KECCAK_SPONGE_WORDS - md->sha3.capacity_words - 1] ^= CONST64(0x8000000000000000);
       keccakf(md->sha3.s);
+      /* store sha3.s[] as little-endian bytes into sha3.sb */
+      for(i = 0; i < SHA3_KECCAK_SPONGE_WORDS; i++) STORE64L(md->sha3.s[i], md->sha3.sb + i * 8);
```
Comment (sjaeckel) [inline src/hashes/sha3.c]: should we probably put the `STORE64L(...)` in braces? 
```
for...{
  STORE64L...
}
```
I more and more get the impression that is better readable...

**Comment by sjaeckel** [comment]
It also works fine for me on `qemu ppc64` and `qemu sparc64`.

Please fix comments, rebase, squash together & merge.

File: src/hashes/sha3.c (hunk 1 above)
Comment (karel-m) [inline src/hashes/sha3.c]: Note that we use `i` twice, the second case is:
```c
  for (idx = 0; idx < outlen; idx++) {
      if (something) {
         /* ... */
         for(i = 0; i < SHA3_KECCAK_SPONGE_WORDS; i++) {
            STORE64L(s[i], sb + i * 8);
         }
         /* ... */
      }
      out[idx] = /* ... */;
   }
```

File: src/hashes/sha3.c (hunk 2 above)
Comment (karel-m) [inline src/hashes/sha3.c]: done

File: src/hashes/sha3.c (hunk 1 above)
Comment (sjaeckel) [inline src/hashes/sha3.c]: dammit, true

File: src/hashes/sha3.c [hunk 3]
```diff
@@ -231,33 +231,27 @@ int sha3_process(hash_state *md, const unsigned char *in, unsigned long inlen)
...
-      for(i = 0; i < SHA3_KECCAK_SPONGE_WORDS; i++) {
-         const ulong32 t1 = (ulong32)(md->sha3.s[i] & CONST64(0xFFFFFFFF));
-         const ulong32 t2 = (ulong32)(md->sha3.s[i] >> 32);
-         STORE32L(t1, md->sha3.sb + i * 8);
-         STORE32L(t2, md->sha3.sb + i * 8 + 4);
-      }
-   }
-#endif
+   /* store sha3.s[] as little-endian bytes into sha3.sb */
+   for(i = 0; i < SHA3_KECCAK_SPONGE_WORDS; i++) STORE64L(md->sha3.s[i], md->sha3.sb + i * 8);
```
Comment (sjaeckel) [inline src/hashes/sha3.c]: \*ping\*

Comment (karel-m) [inline src/hashes/sha3.c]: fixed

**Pull request merged**